wurst/data/Buffer.wurst

```diff
@@ -0,0 +1,602 @@
+				return null
+
+			terminatorIndex = readBuffer.indexOf(TERMINATOR)
+
+			if terminatorIndex < 0
+				// if we haven't found the terminator character in the buffer,
+				// we can be sure that we'll need to return the contents of the
+				// current buffer, and then search in the next chunk
+				value += popAll()
+			else
+				// add the contents up to the terminator character, and then pop
+				// it from the buffer
+				value += readBuffer.substring(0, terminatorIndex)
+				// skip the terminator
+				readBuffer = readBuffer.substring(terminatorIndex + 1)
+		return value
+
+	// returns length amount of characters from the buffer, if there is enough
+	private function popString(int length) returns string
+		if SAFETY_CHECKS_ENABLED and length > MAX_BUFFER_SIZE
+			fail(BufferFailureMode.INTERNAL_ERROR, "trying to pop a string above max size")
+
+		checkDataAvailable()
+		// if there isn't enough data in the readBuffer, load additional chunks
+		while isGood() and readBuffer.length() < length
+			nextChunk()
+
+		if isGood()
+			// if we haven't errored out at this point, it's safe to pop the string
+			let value = readBuffer.substring(0, length)
+			readBuffer = readBuffer.substring(length)
+
+			return value
+		else
+			return null
+
+	// attempts to read the current position as a value type
+	private function popTypeIdentifier() returns ValueType
+		let valueType = valueTypeFromString(popString(1))
+		if valueType == ValueType.INVALID
+			fail(BufferFailureMode.MALFORMED_INPUT, "malformed input")
+		return valueType
+
+	// checks that we are reading the correct type
+	private function checkType(ValueType toCheck) returns boolean
+		let currentType = popTypeIdentifier()
+		// might have malformed input after popping type identifier, so check that first
+		if not isGood()
+			return false
+		if currentType != toCheck
+			fail(BufferFailureMode.INCORRECT_TYPE, "tried to read " + toCheck.toString() + " but was " + currentType.toString())
+			return false
+		return true
+
+	/*
+		Interface implementation
+	*/
+
+	override function writeInt(int value)
+		checkWrite()
+		pushTypeIdentifier(ValueType.INTEGER)
+		pushString(value.toString())
+		pushTerminator()
+
+	override function writeReal(real value)
+		checkWrite()
+		pushTypeIdentifier(ValueType.REAL)
+		pushString(value.toString())
+		pushTerminator()
+
+	override function writeString(string value)
+		if value.length() > MAX_BUFFER_SIZE
+			error("StringBuffer: trying to write string above MAX_BUFFER_SIZE")
+		checkWrite()
+		pushTypeIdentifier(ValueType.STRING)
+		pushString(value.length().toString())
+		pushTerminator()
+		pushString(value)
+
+	override function writeBoolean(bool value)
+		checkWrite()
+		pushTypeIdentifier(ValueType.BOOLEAN)
+		if value
+			pushString("1")
+		else
+			pushString("0")
+
+	private function readIntInternal() returns int
+		return S2I(popUntilTerminator())
+
+	override function readInt() returns int
```

Comment:
Either make read-functions inline friendly or use ternary op to shorten

Reply:
Done